share/spack/qa/setup.sh

```diff
@@ -31,10 +31,10 @@ if [[ "$COVERAGE" == "true" ]]; then
     bashcov=$(realpath ${QA_DIR}/bashcov)
 
     # instrument scripts requiring shell coverage
-    sed -i~ "s@#\!/bin/bash@#\!${bashcov}@" "$SPACK_ROOT/lib/spack/env/cc"
+    sed -i "s@#\!/bin/bash@#\!${bashcov}@" "$SPACK_ROOT/lib/spack/env/cc"
     if [ "$(uname -o)" != "Darwin" ]; then
         # On darwin, #! interpreters must be binaries, so no sbang for bashcov
-        sed -i~ "s@#\!/bin/sh@#\!${bashcov}@"   "$SPACK_ROOT/bin/sbang"
```

Comment:
Doesn't BSD sed require an argument for the inplace flag?

Reply:
MacOS certainly does

Reply:
macOS sed == BSD sed

Reply:
It's not strictly needed, as far as I understand (emphasis mine):

> Edit files in-place, saving backups with the specified extension.  **If a zero-length extension is given, no backup will be saved.**  It is not recommended to give a zero-length extension when in-place editing files, as you risk corruption or partial content in situations where disk space is exhausted, etc.


Also, on MacOS we don't use this script to run tests:

https://github.com/spack/spack/blob/cd91abcf880e61c63c10b0f51be67f482b2149fc/.github/workflows/unit_tests.yaml#L381-L397

as there are also other parts that are not compatible (if I recall correctly we use `realpath` somewhere which is not on Darwin). On Linux having this substitution results in having `~` file accounted for in coverage reports, at least with latest kcov.

Reply:
I'm pretty sure it's needed on macOS (maybe `-I ''` is what they mean by zero-length extension) but as long as we don't run this test on macOS then it doesn't really matter.

Reply:
@adamjstewart @becker33 Ready for another review then :grimacing:

Reply:
Based on https://stackoverflow.com/questions/7573368/in-place-edits-with-sed-on-os-x I would use `-i ''` just in case. This shouldn't change anything on Linux and will allow things to work properly on Darwin. `setup.sh` is called by all of our unit testing scripts, so I assume we use it somewhere on macOS.

Reply:
I'll do the change, but I don't think this is true:

> setup.sh is called by all of our unit testing scripts, so I assume we use it somewhere on macOS.

The only place where we run tests on a MacOS runner is in the snippet above and:
```
% share/spack/qa/setup.sh 
share/spack/qa/setup.sh: line 14: realpath: command not found
share/spack/qa/setup.sh: line 17: /share/spack/setup-env.sh: No such file or directory
```

Reply:
We should probably not use realpath lol. I wrote an entire gist for this cuz I could never remember the best way to find an absolute path: https://gist.github.com/adamjstewart/c7629ed2ad89ed25513a82d3cb5d774e. But this can be fixed another day. As long as we aren't introducing additional Linux-isms.

Reply:
@adamjstewart Reverted since using a zero-length argument with a space seems to cause failures on linux :slightly_frowning_face:

Reply:
Bleh, according to https://stackoverflow.com/questions/5694228/sed-in-place-flag-that-works-both-on-mac-bsd-and-linux it's just not possible to support both. Let's go with what you have now. If we ever need macOS support we can add an if-statement.

Reply:
```
-i[SUFFIX], --in-place[=SUFFIX]

      edit files in place (makes backup if SUFFIX supplied)
```